File: app/src/com/example/portalframe/AmbientColor.java
package com.example.portalframe;

import android.graphics.Bitmap;
import android.graphics.Color;

/**
 * Extracts a single vibrant "mood" color from a photo so the frame's edges can
 * glow with it — a living, color-matched frame (think Ambilight for photos).
 *
 * Pure Java, no dependencies: downscale to a tiny grid and take a
 * saturation-weighted average, then normalise so even muted photos give a clear
 * but tasteful tint. Runs in well under a frame at slide-change time.
 */
final class AmbientColor {

    private static final int GRID = 32;

    private AmbientColor() {}

    /** An opaque, pleasantly saturated color representing the photo, or null. */
    static Integer extract(Bitmap src) {
        if (src == null || src.getWidth() < 2 || src.getHeight() < 2) {
            return null;
        }
        Bitmap s = null;
        try {
            s = Bitmap.createScaledBitmap(src, GRID, GRID, true);
            int[] px = new int[GRID * GRID];
            s.getPixels(px, 0, GRID, 0, 0, GRID, GRID);
            double rw = 0, gw = 0, bw = 0, wsum = 0;
            float[] hsv = new float[3];
            for (int p : px) {
                int r = (p >> 16) & 0xff, g = (p >> 8) & 0xff, b = p & 0xff;
                Color.RGBToHSV(r, g, b, hsv);
                // Favour colorful, mid-bright pixels; ignore near-black/near-white.
                float sat = hsv[1], val = hsv[2];
                if (val < 0.15f || val > 0.97f) {
                    continue;
                }
                double w = sat * sat * (0.4 + 0.6 * val);
                rw += r * w;
                gw += g * w;
                bw += b * w;
                wsum += w;
            }
            if (wsum < 1e-3) {
                return null; // a flat / monochrome photo — no glow
            }
            int r = (int) Math.round(rw / wsum);
            int g = (int) Math.round(gw / wsum);
            int b = (int) Math.round(bw / wsum);
            // Normalise: lift saturation a touch and cap brightness so the glow reads
            // as color without washing out.
            Color.RGBToHSV(r, g, b, hsv);
            hsv[1] = Math.min(1f, hsv[1] * 1.25f + 0.10f);
            hsv[2] = Math.min(0.85f, Math.max(0.45f, hsv[2]));
            return Color.HSVToColor(hsv);
        } catch (Throwable t) {
            return null;
        } finally {
            if (s != null && s != src) {
                s.recycle();
            }
        }
    }
}

File: app/src/com/portalhacks/frame/AmbientColor.kt
package com.portalhacks.frame

import android.graphics.Bitmap
import android.graphics.Color

/**
 * Extracts a single vibrant "mood" color from a photo so the frame's edges can
 * glow with it — a living, color-matched frame (think Ambilight for photos).
 *
 * Pure Kotlin, no dependencies: downscale to a tiny grid and take a
 * saturation-weighted average, then normalise so even muted photos give a clear
 * but tasteful tint. Runs in well under a frame at slide-change time.
 */
internal object AmbientColor {

    private const val GRID = 32

    /** An opaque, pleasantly saturated color representing the photo, or null. */
    @JvmStatic
    fun extract(src: Bitmap?): Int? {
        if (src == null || src.width < 2 || src.height < 2) {
            return null
        }
        var s: Bitmap? = null
        try {
            s = Bitmap.createScaledBitmap(src, GRID, GRID, true)
            val px = IntArray(GRID * GRID)
            s.getPixels(px, 0, GRID, 0, 0, GRID, GRID)
            var rw = 0.0
            var gw = 0.0
            var bw = 0.0
            var wsum = 0.0
            val hsv = FloatArray(3)
            for (p in px) {
                val r = (p shr 16) and 0xff
                val g = (p shr 8) and 0xff
                val b = p and 0xff
                Color.RGBToHSV(r, g, b, hsv)
                // Favour colorful, mid-bright pixels; ignore near-black/near-white.
                val sat = hsv[1]
                val v = hsv[2]
                if (v < 0.15f || v > 0.97f) {
                    continue
                }
                val w = sat * sat * (0.4 + 0.6 * v)
                rw += r * w
                gw += g * w
                bw += b * w
                wsum += w
            }
            if (wsum < 1e-3) {
                return null // a flat / monochrome photo — no glow
            }
            val r = Math.round(rw / wsum).toInt()
            val g = Math.round(gw / wsum).toInt()
            val b = Math.round(bw / wsum).toInt()
            // Normalise: lift saturation a touch and cap brightness so the glow reads
            // as color without washing out.
            Color.RGBToHSV(r, g, b, hsv)
            hsv[1] = minOf(1f, hsv[1] * 1.25f + 0.10f)
            hsv[2] = minOf(0.85f, maxOf(0.45f, hsv[2]))
            return Color.HSVToColor(hsv)
        } catch (t: Throwable) {
            return null
        } finally {
            if (s != null && s !== src) {
                s.recycle()
            }
        }
    }
}
